### PR TITLE
CMake build system for git

### PR DESCRIPTION
**Pull request by SibiSiddharthan**
This is an attempt to build Git using CMake. CMake is cross-platform build generator which works well on a variety of platforms(primarily Linux and Windows). Using CMake we can check whether
certain headers exist, certain functions exist, the required libraries are present and configure the build accordingly.  Using CMake we can also build and test Git out of source, resulting in a clean source tree.

Tested platforms

Ubuntu 18.04
GCC 7.4
Clang 8.0.1

Windows
MinGW GCC 9.2
Clang 9
Visual Studio 2015,2017,2019

Changes:
1) The CMake script has been relocated to contrib/buildsystems
2) The CMake script parses the Makefile for the sources.
LIB_OBJS
BUILTIN_OBJS
XDIFF_OBJS
VCSSVN_OBJS
TEST_BUILTINS_OBJS
SCRIPT_SH
SCRIPT_PERL
3) Philip suggested to change the error message if sh/bash was not
found on windows.
4) CMake now tests for ICONV_OMITS_BOM, NO_ST_BLOCKS_IN_STRUCT_STAT
5) Renamed the variable test_helper_sources to test-tool_SOURCES [PATCH 4/8]
to be consistent with the naming of source variables.

Changes v2:
Changes 1,2,4 have been rebased to PATCH 01/xx
CMake uses GIT-VERSION-GEN to get the version of Git
Fixed a bug where a Windows user can pose as Linux user and vice versa. [PATCH 6/8]

Changes v3:
Patch changes are moved from the commit messages and are placed here.
Code inside check_c_source_(compiles/runs) have been formatted according to git coding guidelines. [PATCH 1/8]
The CMake script parses the Makefile for SCRIPT_LIB also. [PATCH 2/8]
The CMake script globs templates, po files. Logic has been added to place the template files in their respective directories instead of hard-coding them.  [PATCH 2/8]

Changes v4:
Removed EXE_EXTENSION conditional stuff using CMAKE_EXECUTABLE_SUFFIX [PATCH 4/8]
There was an issue in build pipelines where CMake was not able to find the correct iconv library (caused by an update that installed PostgreSQL), so we need to manually set the location of the iconv library and its includes. This issue is extremely rare and is specific to the implementation of FindIconv.cmake. Other libraries are unaffected. [PATCH 8/8]

### CODE REVIEW

**Comment by dscho** [comment]
Would you mind aligning the style of the commit messages more with the existing ones? For example,

>  CMake build system for git on Linux 

should probably read more like

>  Introduce support for configuring on Linux via CMake

And instead of

> CMake 3.14 is chosen because it offers a cross platform way of creating
hardlinks.

the commit message should be a lot more verbose. Personally, I would do something like this:

```
Introduce CMake support for configuring Git on Linux

At the moment, the recommended way to configure Git's builds is to
simply run `make`. If that does not work, the recommended strategy is to
look at the top of the `Makefile` to see whether any "Makefile knob" has
to be turned on/off, e.g. `make NO_OPENSSL=YesPlease`.

Alternatively, Git also has an `autoconf` setup which allows configuring
builds via `./configure [<option>...]`.

Both of these options are fine if the developer works on Unix or Linux.
But on Windows, we have to jump through hoops to configure a build
(read: we force the user to install a full Git for Windows SDK, which
occupies around two gigabytes (!) on disk and downloads about three
quarters of a gigabyte worth of Git objects).

To make this a little less awkward, the Git for Windows project offers
the `vs/master` branch which has a full Visual Studio solution generated
and committed. This branch can therefore be used to tinker with Git in
Visual Studio _without_ having to download the full Git for Windows SDK.
Unfortunatly, that branch is auto-generated from Git for Windows'
`master`. If a developer wants to tinker, say, with `pu`, they are out
of luck.

CMake was invented to make this sort of problem go away, by providing a
more standardized, cross-platform way to configure builds.

With a working support CMake, developers on Windows need only install
CMake, configure their build, load the generated Visual Studio solution
and immediately start modifying the code and build their own version of
Git. Likewise, developers on other platforms can use the convenient GUI
tools provided by CMake to configure their build.

So let's start building CMake support for Git.

This is only the first step, and to make it easier to review, it only
allows for configuring builds on the platform that is easiest to
configure for: Linux.

Note: earlier endeavors on the Git mailing list to introduce CMake ended
up in dead water. The primary reason for that was that reviewers
_expected_ CMake support to fall out of maintenance, unless the
contributor would promise to keep an eye on keeping CMake support up to
date. However, in the meantime, support for automated testing has been
introduced in Git's source code, and a later patch will modify the
(still experimental) GitHub workflow to continually verify that CMake
support is still complete. That will make maintenance reasonably easy.

Note: this patch asks for the minimum version v3.14 of CMake (which is
not all that old as of time of writing) because that is the first
version to offer a platform-independent way to generate hardlinks as
part of the build. This is needed to generate all those hardlinks for
the built-in commands of Git.
```

Ideally, _all_ of these commit messages will be equally verbose and convincing.

(And their "onelines", i.e. the first lines of their commit messages, won't read "cmake: update" for the most part 😄 )

**Comment by dscho** [comment]
To add to my previous comment: The commit message

```
 cmake: update

Added generating shell, perl and python scripts with CMake
```

would probably read a lot better like this:

```
cmake: generate the shell/Perl/Python scripts

This patch implements the placeholder substitution to generate, say,
`git-pull-request` from `git-pull-request.sh`.

Note that we reimplement in a CMake-native way the `sed` invocation that
is defined as `cmd_munge_script` in the `Makefile`.
```

Following this example, the remainder of the commit messages are probably eager to be enhanced in the same manner 😄

**Comment by SibiSiddharthan** [comment]
Will reword the comments

**Comment by SibiSiddharthan** [comment]
Have reworded the commit messages, explained what each patch does.

**Comment by SibiSiddharthan** [comment]
@dscho shall I submit for a PR?

**Comment by dscho** [comment]
Could you rephrase the onelines so that they are in the imperative form? "installation support" are two nouns whereas "support installing Git" would be a full, imperative sentence.

Also, it would be interesting to the occasional reader to learn more about the background, e.g. The commit message of the commit that introduces support for CTest should probably describe how to run CTest and how its output looks like.

The idea is to take the reader by their hand, where they are, and guide them along the story arc right to the glorious conclusion, when CMake support is complete, and everyone lives happily ever after.

Remember: previous attempts to introduce CMake support failed. This is not an audience that wants it. Your commit messages need to convey why they actually probably do want it, and that it's not actually hard to have/maintain anymore.

**Comment by SibiSiddharthan** [comment]
@dscho
Have reworded the 2 commit messages pointed out.
Have also rephrased the cover letter.
Waiting for your review

Thank You,
Sibi Siddharthan

**Comment by dscho** [comment]
Don't other onelines use the "ing" form instead of an imperative? Aren't there more opportunities to provide useful background?

**Comment by SibiSiddharthan** [comment]
@dscho
I have provided as much as background as possible for the commit messages, if more detail has to be provided let me know.

In the CI it looks like CMake is broken, it is not able to find libcurl, whereas previously there wasn't any problem. Yesterday my builds ran fine, today I am just rewording the commits builds are failing.

**Comment by dscho** [comment]
> I have provided as much as background as possible for the commit messages, if more detail has to be provided let me know.

@SibiSiddharthan I just thought that e.g. the comment on `invalidcontinue.obj` was a bit sparse. If I was not familiar with the issue, I would love to see _at least_ a reference to 4b623d80f735 (MSVC: link in invalidcontinue.obj for better POSIX compatibility, 2014-03-29), ideally accompanied by a short summary ("This prevents the process from aborting instead of setting `errno=EINVAL` on invalid parameters to msvcrt functions" or some such).

> In the CI it looks like CMake is broken, it is not able to find libcurl, whereas previously there wasn't any problem. Yesterday my builds ran fine, today I am just rewording the commits builds are failing.

Hmm. This unfortunately happens from time to time. Last time, `vcpkg` changed the location or name of the library, if I recall correctly.

> It seems they have upgraded to CMake 3.17.1, I have also locally tested with this version and it does not find the curl library when CURLConfig.cmake is present. This is buggy behaviour, but it can be disabled with -DCURL_NO_CURL_CMAKE=ON

Excellent find!

**Comment by dscho** [comment]
As you are targeting `next`, and as `bloom.c` has been included in `next` as of yesterday night, I think this needs to be added to the `CMakeLists.txt` file, now, too.

**Comment by dscho** [comment]
BTW I would change the prefix `github actions:` to `ci:` to make it shorter. And please make sure that you spell GitHub consistently. There is no "th" sound in GitHub, therefore it is incorrect to use a lower-case "h". And an all-lowercase "github" is completely wrong ;-)

**Comment by SibiSiddharthan** [comment]
@dscho 
Have referenced the commit 4b623d8 in the respective commit.
GitHub everywhere :-)
Have made changes according to the latest change in the next branch.

Any changes that I have to make before a submit.

Thank You,
Sibi Siddharthan

**Comment by dscho** [comment]
I guess it's time to submit, yes.

**Comment by SibiSiddharthan** [comment]
@dscho 
It seems there is an error when I submit , something related to workflow permissions.

**Comment by dscho** [comment]
Ouch. Right. GitGitGadget cannot push that tag because GitHub apps can't change workflows. I suspect that the mails were sent, though.

**Comment by dscho** [comment]
Yep: https://lore.kernel.org/git/pull.614.git.1587700897.gitgitgadget@gmail.com/T/#t

**Comment by dscho** [comment]
@webstech FYI the token problem is back :-(

I'll hopefully get to it in ~4h.

**Comment by dscho** [comment]
@SibiSiddharthan okay, as per gitgitgadget/gitgitgadget#226, this bug should be solved. Actually, even before that PR is merged, the bug should be solved, I explained in that PR what reaction work I performed to address this problem.

**Comment by SibiSiddharthan** [comment]
@dscho 
Thanks for the help.
Does this mean that I have to submit again after this 
"https://github.com/gitgitgadget/gitgitgadget/pull/226"
has been merged?

**Comment by dscho** [comment]
No, no need to re-`/submit`. as I pointed out, the mails made it to the list :-)

**Comment by SibiSiddharthan** [comment]
@dscho 
When I changed the vs-test step to match windows-test step without the CI specific blocks(Windows absolute paths) the tests ran 6 times faster.
With the CI specific blocks in place there was no much difference but when you look at the individual test times the vs-test(*) were about 2-4 mins longer than the windows-test ones consistently.
I can guarantee this.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -97,12 +97,10 @@ jobs:
...
     - name: generate Visual Studio solution
-      shell: powershell
+      shell: cmd
       run: |
-        & .\git-sdk-64-minimal\usr\bin\bash.exe -lc @"
-          cmake ./contrib/buildsystems/ -DCMAKE_PREFIX_PATH=`pwd`/compat/vcbuild/vcpkg/installed/x64-windows -DMSGFMT_EXE=./git-sdk-64-minimal/mingw64/bin/msgfmt.exe -DPERL_TESTS=OFF -DPYTHON_TESTS=OFF -DCURL_NO_CURL_CMAKE=ON
-        "@
-        if (!$?) { exit(1) }
+        & 
+        cmake %PWD%/contrib/buildsystems/ -DCMAKE_PREFIX_PATH=%PWD%/compat/vcbuild/vcpkg/installed/x64-windows -DMSGFMT_EXE=%PWD%/git-sdk-64-minimal/mingw64/bin/msgfmt.exe -DPERL_TESTS=OFF -DPYTHON_TESTS=OFF -DCURL_NO_CURL_CMAKE=ON
```
Comment (dscho) [inline .github/workflows/main.yml:103]: If you want to, you can also use `shell: bash`, which uses a Git for Windows Bash on the build agents.

Comment (SibiSiddharthan) [reply .github/workflows/main.yml:103]: I having trouble specifying the absolute directory for CMAKE_PREFIX_PATH, will do

Comment (dscho) [reply .github/workflows/main.yml]: I think it expands, but PWD is simply not a thing in `cmd.exe`.

**Comment by SibiSiddharthan** [comment]
@dscho git-skd-64-minimal is no longer downloading

**Comment by dscho** [comment]
> @dscho git-sdk-64-minimal is no longer downloading

Indeed. The problem is that it blew right through my quota on Azure.

I opened https://github.com/git-for-windows/git/pull/2622 to fix this. We will most likely need a real, real quick fix for git.git, too. I'll open a GitGitGadget PR for this.

**Comment by dscho** [comment]
> I'll open a GitGitGadget PR for this.

Here you go: https://github.com/gitgitgadget/git/pull/632

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -125,9 +124,9 @@ jobs:
         nr: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
     steps:
     - uses: actions/checkout@v1
-    - name: download git-64-portable
+    - name: download git-sdk-64-minimal
       shell: bash
-      run: a=git-64-portable && mkdir -p $a && curl -# https://wingit.blob.core.windows.net/ci-artifacts/$a.tar.xz | tar -C $a -xJf -
+      run: a=git-sdk-64-minimal && mkdir -p $a && curl -# https://wingit.blob.core.windows.net/ci-artifacts/$a.tar.xz | tar -C $a -xJf -
```
Comment (dscho) [inline .github/workflows/main.yml:129]: To avoid hammering our `ci-artifacts` container, in https://github.com/gitgitgadget/git/pull/632 I now also use `git-sdk-64-minimal`. While this does not test the intended workflow (when building with Visual Studio, we do _not_ want to require users to download Git for Windows' SDK just to run the tests, we want the tests to run in a regular Git Bash), it needs to be good enough for now.

Comment (dscho) [reply .github/workflows/main.yml:129]: (In other words, I do not ask you to change anything here.)

**Comment by SibiSiddharthan** [comment]
On Wed, May 13, 2020 at 2:59 AM gitgitgadget[bot]
<notifications@github.com> wrote:
>
> @gitgitgadget[bot] commented on this pull request.
>
> ________________________________
>
> In .github/workflows/main.yml:
>
> > @@ -80,13 +80,6 @@ jobs:
>      - name: download git-sdk-64-minimal
>
> On the Git mailing list, Junio C Hamano wrote (reply to this):
>
> "Sibi Siddharthan via GitGitGadget" <gitgitgadget@gmail.com> writes:
>
> > From: Sibi Siddharthan <sibisiddharthan.github@gmail.com>
> > Subject: Re: [PATCH v2 11/11] ci: modification of main.yml to use cmake for vs-build job
>
> Yay!
>
> > This patch modifies .github/workflows/main.yml to use CMake for
> > Visual Studio builds.
> > Modified the vs-test step to match windows-test step. This speeds
>
> Teach .github/workflows/main.yml to use Cmake for VS builds.
> Modify vs-test step to match windows-test step to speed up
> the vs-test.
>
> > Changes:
> > The CMake script has been relocated to contib/buildsystems, so point
> > to the CMakeLists.txt in the invocation commands.
>

Will remove this.

> This does not belong to the log message for this (or any for that
> matter) commit, no?  Up to this point, nothing in main.yml used
> CMake script from anywhere, and "has been relocated" is irrelevant.
>
> And if we add CMake script to contrib/buildsystems/ from the
> beginning in [01/11], there won't be any "relocation" the readers of
> the log message of this step need to know about.
>
> > The generation command now uses the absolute paths for the generation
> > step.
>
> "now uses"?  Is that something the readers of the log message of
> this step need to know about, or is it "in contrast to an earlier
> attempt to add CMake script"?  If the latter, it only confuses the
> readers of our history, as most of them will not even know about an
> earlier attempt.

Will also remove this.

>
> Thanks.
>

Thank You,
Sibi Siddharthan
